### ISSUE TITLE
NVDA fails to switch to a proper language-specific voice when the typing language is altered via ALT+Shift

### ISSUE THREAD

**Issue by amirsol81** [nvaccess/nvda#7924]
### Steps to reproduce:
1. Make sure Automatic Language Switching and Automatic Dialect Switching are enabled in the Voice Settings dialog
2. Make sure you're using a TTS engine, like Eloquence/eSpeak/Vocalizer, which supports multiple languages
3. Open an MS Word or Notepad document, start typing, alter the language by pressing ALT+Shift, and keep typing
4. You'll notice that NVDA keeps using its default TTS voice and fails to switch to a proper language-specific TTS voice as the typing language is altered.

This issue is quite problematic and pronounced in MS Word which comes with enhanced support for editing and proof-reading multiple languages via its Review ribbon. So even when NVDA properly switches to a new language-specific voice as the language of the text changes upon pressing Up/Down or in automatic reading, you won't be able to get NVDA to automatically pick the proper language-specific voice for typing within the new block of text unless, of course, you alter the voice language manually by bringing up the Voice Settings dialog which is time-consuming.
Expected behavior: As the typing language is altered or as MS Word detects a new language, NVDA should switch to a proper voice for that language which can be used for typing in that language.

Test environment: Windows 10 Pro 64-Bit, MS Word 2016, Notepad, NVDA 2017.4, Eloquence/Vocalizer with multiple languages installed.

**Comment by feerrenrut** [comment]
I can reproduce this. It's worth noting that this only affects "speak typed words" or "speak typed characters". When reading back the words with `ctl+left arrow` the voices change as expected.

**Comment by amirsol81** [comment]
Yes. And that's why I put emphasis on the typing aspect of the issue.
This is a problem which makes typing in multilingual environments
rather painful.
Thanks.

**Comment by amirsol81** [comment]
This is still duplicable with NVDA alpha-20005,14c2e2ec and Windows 10 Pro 64-Bit Version 2004 (OS Build 19041.207).

**Comment by JulienCochuyt** [comment]
I agree this situation is an issue.
Still, when attempting to fix it, please make it somehow configurable: I'm personally commonly switching between 3 input layout across 2 input languages, while reading and typing text in 4 different languages.
Automatic speech language switch while I switch layouts would surely help me in a few situations, but would also be the heck of a pain in many others.

**Comment by amirsol81** [comment]
@JulienCochuyt Same here. But if the fix gets implemented properly, we'll have zero issues to deal with. That is, this works perfectly in JAWS, so there's no reason to assume that the fix will make life difficult for NVDA users like you and I.

**Comment by amirsol81** [comment]
@feerrenrut @seanbudd Any chance of taking care of this issue? It does make life a whole lot easier for those of us who should work in multilingual or bilingual environments. It also helps a lot with language learning - the way, say, JAWS does it.

**Comment by amirsol81** [comment]
@feerrenrut @seanbudd Forgot to mention that the issue is still duplicable with NVDA Alpha 26134 and Windows 11 22H2.

**Comment by Qchristensen** [comment]
For those reporting this, does it happen with UIA enabled?

And has anyone looked at it recently?  It's still being reported.

**Comment by amirsol81** [comment]
@Qchristensen I can still duplicate this with MS Word 2021, and I'm using UIA. However, since this is language/TTS-related, the use of UIA isn't related as, for instance and unlike JAWS, NVDA has the same issue in Notepad or other text areas. As such, this is more of a language/synth issue rather than a Word/Office issue though it's much more pronounced in Office applications.

**Comment by amirsol81** [comment]
@Qchristensen @seanbudd @josephsl  Any chance of also taking a look at this? It's still duplicable with NVDA Alpha 32997 and 2024.3 betas.

**Comment by josephsl** [comment]
Hi,

This may require looking at character/input processing and related modules. One idea would have been Unicode range detection and use appropriate speech engine/voice combination. However, note that what NVDA sees from the keyboard (a bit low level) versus what Word and other aps interpret may vary, making the solution quite complex.

Thanks.

**Comment by amirsol81** [comment]
@josephsl Thanks. I brought it up, once again, because JAWS does it flawlessly and I thought maybe NVDA can also handle the situation gracefully.
